Implement `FromRedisValue` for `Box<[T]>` and `Arc<[T]>`

JOT85 opened the pull request.

nihohit commented:
please add tests.

JOT85 commented on redis/src/types.rs:
```diff
@@ -1229,22 +1229,44 @@ impl FromRedisValue for String {
 
+impl<T: FromRedisValue> FromRedisValue for Box<[T]> {
+    fn from_redis_value(v: &Value) -> RedisResult<Box<[T]>> {
+        match v {
```
The problem with that it would then have an error message that isn't quite right, it wouldn't be `Conversion to Box<{}> failed`, but would instead be the same error message as when converting to a `Vec`.

jaymell replied:
Honestly I'm wondering how much benefit this impl offers, but at a minimum I think we should avoid the duplicated code somewhere along the lines of what @nihohit suggested.

nihohit reviewed (approved):
LGTM

JOT85 commented:
@nihohit would you like me to make those changes to this branch? It looks like #961 might have superseded this?

nihohit commented:
@JOT85 sorry, I wasn't sure if you're still interested in this, so I moved forward and integrated your change in #961.
If you still want, you're more than welcome to update your PR either with my changes, or your version - you created this first, and if you want this contribution under your name, it's rightly yours.

JOT85 commented:
Sorry it took so long to get round to making that change!

I switched the impls to a macro, and also added an impl for `Arc<[T]>`.

The differences between the implementation I used are:

- The macro generates the entire impl, not just the function.
- The error message includes the collection type, not just the inner type, for example it's `Conversion to Box<[u32]> failed` instead of `Conversion to u32 failed`.

nihohit reviewed (approved):
@jaymell LGTM.
@JOT85 The test code has quite a bit of repetition, it would be nice if some of that can be reduced.

jaymell commented:
> @jaymell LGTM. @JOT85 The test code has quite a bit of repetition, it would be nice if some of that can be reduced.

Thanks! -- I'll go ahead and merge since we've been sitting on this awhile.
